README.md
# Monitor Transcend SD card health and status
Get S.M.A.R.T data for Transcend SD card.

-------------------------
The various statuses of MicroSD products are located in the SMART block of the device. 
By reading and analyzing the SMART block, you can get the product name, serial number, firmware version, Erase count, health, etc. of the device, so as to monitor and find the 
device early potential problems. This document provides a way to use the the test tool that get SMART block to enable SMART command via the native MicroSD port on embedded linux such as Raspberry Pi or Wiren Board.

Hardware requirements
-------------------------
* Embedded Linux with native SD interface (/dev/mmcblk*)
* Transcend Micro SD 430T / 450I

Usage
-------------------------

```
$ sudo ./transcend-sdcard-health –h
``` 
1. Get SMART buffer by following command.
```
$ sudo ./transcend-sdcard-health smtbuffer /dev/mmcblk0
``` 
2. Get SMART information by following command.
```
$ sudo ./transcend-sdcard-health smart /dev/mmcblk0
``` 
3. Get Card life by following command.
```
$ sudo ./transcend-sdcard-health health /dev/mmcblk0
``` 

Command Execution
-------------------------
1. To dump SMART raw data</br></br>
![Buffer](https://github.com/transcend-information/RaspberryPi-SDcard-SMARTQuery/blob/main/smtbuffer.png)

2. To show SMART info</br></br>
![SMART](https://github.com/transcend-information/RaspberryPi-SDcard-SMARTQuery/blob/main/smart.png)

3. To show card life(Heath)</br></br>
![Heath](https://github.com/transcend-information/RaspberryPi-SDcard-SMARTQuery/blob/main/health.png)


debian/changelog
transcend-sdcard-health (1.0) stable; urgency=low

  * Initial release. 

 -- Evgeny Boger <boger@wirenboard.com>  Sun, 31 Jul 2022 01:05:46 +0300


mmc.h
/*
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License v2 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 021110-1307, USA.
 *
 * Modified to add field firmware update support,
 * those modifications are Copyright (c) 2016 SanDisk Corp.
 */

#include <linux/mmc/ioctl.h>

/* From kernel linux/mmc/core.h */
#define MMC_RSP_PRESENT	(1 << 0)
#define MMC_RSP_136	(1 << 1)		/* 136 bit response */
#define MMC_RSP_CRC	(1 << 2)		/* expect valid crc */
#define MMC_RSP_BUSY	(1 << 3)		/* card may send busy */
#define MMC_RSP_OPCODE	(1 << 4)		/* response contains opcode */

#define MMC_CMD_AC	(0 << 5)
#define MMC_CMD_ADTC	(1 << 5)

#define MMC_RSP_SPI_S1	(1 << 7)		/* one status byte */
#define MMC_RSP_SPI_BUSY (1 << 10)		/* card may send busy */

#define MMC_RSP_SPI_R1	(MMC_RSP_SPI_S1)
#define MMC_RSP_SPI_R1B	(MMC_RSP_SPI_S1|MMC_RSP_SPI_BUSY)

#define MMC_RSP_R1	(MMC_RSP_PRESENT|MMC_RSP_CRC|MMC_RSP_OPCODE)
#define MMC_RSP_R1B	(MMC_RSP_PRESENT|MMC_RSP_CRC|MMC_RSP_OPCODE|MMC_RSP_BUSY)

/* SD SMART */
#define MMC_RSP_R2 (MMC_RSP_PRESENT|MMC_RSP_136|MMC_RSP_CRC)
#define MMC_CMD_BCR (3 << 5)
#define SD_CMD 56
#define SD_SMT_BLOCK_SIZE 512 /* block size of smart data*/
#define TYPE_SMART (0x01)
#define TYPE_HEALTH (0x02)


mmc_cmds.c
/*
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License v2 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 021110-1307, USA.
 *
 * Modified to add field firmware update support,
 * those modifications are Copyright (c) 2016 SanDisk Corp.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <stdint.h>
#include <assert.h>
#include <linux/fs.h> /* for BLKGETSIZE */
#include "mmc.h"
#include "mmc_cmds.h"

enum rpmb_op_type {
	MMC_RPMB_WRITE_KEY = 0x01,
	MMC_RPMB_READ_CNT  = 0x02,
	MMC_RPMB_WRITE     = 0x03,
	MMC_RPMB_READ      = 0x04,

	/* For internal usage only, do not use it directly */
	MMC_RPMB_READ_RESP = 0x05
};

struct rpmb_frame {
	u_int8_t  stuff[196];
	u_int8_t  key_mac[32];
	u_int8_t  data[256];
	u_int8_t  nonce[16];
	u_int32_t write_counter;
	u_int16_t addr;
	u_int16_t block_count;
	u_int16_t result;
	u_int16_t req_resp;
};

static inline void set_single_cmd(struct mmc_ioc_cmd *ioc, __u32 opcode,
				  int write_flag, unsigned int blocks)
{
	ioc->opcode = opcode;
	ioc->write_flag = write_flag;
	ioc->arg = 0x0;
	ioc->blksz = 512;
	ioc->blocks = blocks;
	ioc->flags = MMC_RSP_SPI_R1 | MMC_RSP_R1 | MMC_CMD_ADTC;
}

int do_SMART_buffer_dump(int nargs, char **argv)/* Show SMART 512-byte buffer by using CMD56 */
{
	int argCmd56 = 0x110005F9;
	char data_buff[SD_SMT_BLOCK_SIZE];
	int fd, ret;
	char *device;
	device = argv[nargs-1];
	fd = open(device, O_RDWR);
	if (fd < 0) {
		perror("open file fail");
		exit(1);
	}
	ret = CMD56_data_in(fd, argCmd56 , data_buff);
	if (ret) {
	fprintf(stderr, "CMD56 function fail, %s\n", device);
	exit(1);
	}
	dump_smart_data(data_buff); 	
	return ret;
}

int show_SMART_info(int nargs, char **argv) /* Show SMART info (ex: Speed class/ Erase count / FW version, etc...)  */
{
	int argCmd56 = 0x110005F9;
	char data_buff[SD_SMT_BLOCK_SIZE];
	int fd, ret;
	char *device;
	device = argv[nargs-1];
	fd = open(device, O_RDWR);
	if (fd < 0) {
		perror("open file fail");
		exit(1);
	}
	ret = CMD56_data_in(fd, argCmd56 , data_buff);
	if (ret) {
	fprintf(stderr, "CMD56 function fail, %s\n", device);
	exit(1);
	}
	is_transcend_card(data_buff, TYPE_SMART); /* Only support microSDXC430T and microSDXC450I */
	return ret;
}

int show_Health_info(int nargs, char **argv) /* Show Health */
{
	int argCmd56 = 0x110005F9;
	char data_buff[SD_SMT_BLOCK_SIZE];
	int fd, ret;
	char *device;
	device = argv[nargs-1];
	fd = open(device, O_RDWR);
	if (fd < 0) {
		perror("open file fail");
		exit(1);
	}
	//Get one 512-byte data block by CMD56 function
	ret = CMD56_data_in(fd, argCmd56 , data_buff);
	if (ret) {
	fprintf(stderr, "CMD56 function fail, %s\n", device);
	exit(1);
	}
	is_transcend_card(data_buff, TYPE_HEALTH); //Transcend SD card only
	return ret;
}

int CMD56_data_in(int fd, int argCmd56, char *block_data_buff)
{
	int ret = 0;
	struct mmc_ioc_cmd ioc_data;
	memset(&ioc_data, 0, sizeof(ioc_data));
	memset(block_data_buff, 0, sizeof(__u8) * SD_SMT_BLOCK_SIZE);
	ioc_data.arg = argCmd56;
	ioc_data.write_flag = 0;
	ioc_data.blocks = 1;
	ioc_data.blksz = SD_SMT_BLOCK_SIZE;
	ioc_data.opcode = SD_CMD;
	ioc_data.flags = MMC_RSP_SPI_R1 | MMC_RSP_R1 | MMC_CMD_ADTC;
	mmc_ioc_cmd_set_data(ioc_data, block_data_buff);
	ret = ioctl(fd, MMC_IOC_CMD, &ioc_data);
	if (ret)
		perror("ioctl fail");
	return ret;
}
void dump_smart_data(char *block_data_buff)
{
	int count=0;
	printf("SMART data block:");
	while( count < SD_SMT_BLOCK_SIZE) {
		if(count % 16 == 0)
		printf("\n%03d: ", count);
		printf("%02x ", block_data_buff[count]);
		count++;
	}
	printf("\n");
	
	return;
}

void is_transcend_card(char *block_data_buff, char function)
{
	if(block_data_buff[0] == 84 && block_data_buff[1] == 114 && block_data_buff[2] == 97 && block_data_buff[3] == 110 && block_data_buff[4] == 115 
		&& block_data_buff[5] == 99 && block_data_buff[6] == 101 && block_data_buff[7] == 110 && block_data_buff[8] == 100)//TRANSCEND
	{
		char controllerChar[9];
		strncpy(controllerChar, block_data_buff + 88, 8);
		if(strstr(controllerChar, "SM2706"))
		{
			if(function == TYPE_SMART)
			{
				parsing_SMART_info(block_data_buff);
			}
			else if(function == TYPE_HEALTH)
			{
				parsing_Health_info(block_data_buff);
			}
		}
		else
		{
			printf("Not Support.\n");
		}
	}
	else
	{
		printf("Not Support.\n");
	}
	return;
}

void parsing_SMART_info(char *block_data_buff) /* parsing SMART 512-byte array */ 
{
	char value[64];
	
	char *card_marker = grabString(block_data_buff, 0, 16);
	sprintf(value, "Card Marker(0x00):\t\t\t%s", card_marker);
	printf("\n%s", value);
	free(card_marker);
	
	char *bus_width = grabHex(block_data_buff, 16, 1);
	if (bus_width[0] == 0x00){
		sprintf(value, "Bus Width(0x10):\t\t\t1 bit Width");
	}
	else if (bus_width[0] == 0x10){
		sprintf(value, "Bus Width(0x10):\t\t\t4 bit Width");
	}
	printf("\n%s", value);
	free(bus_width);
	
	char *secure_mode = grabHex(block_data_buff, 17, 1);
	if (secure_mode[0] == 0x00){
		sprintf(value, "Secured Mode(0x11):\t\t\tNot in the secured mode");
	}
	else if (secure_mode[0] == 0x01){
		sprintf(value, "Secured Mode(0x11):\t\t\tIn secured mode");
	}
	printf("\n%s", value);
	free(secure_mode);
	
	char *speed_class = grabHex(block_data_buff, 18, 1);
	if (speed_class[0] == 0x00){
		sprintf(value, "Speed Class(0x12):\t\t\tClass 0");
	}
	else if (speed_class[0] == 0x01){
		sprintf(value, "Speed Class(0x12):\t\t\tClass 2");
	}
	else if (speed_class[0] == 0x02){
		sprintf(value, "Speed Class(0x12):\t\t\tClass 4");
	}
	else if (speed_class[0] == 0x03){
		sprintf(value, "Speed Class(0x12):\t\t\tClass 6");
	}
	else if (speed_class[0] == 0x04){
		sprintf(value, "Speed Class(0x12):\t\t\tClass 10");
	}
	printf("\n%s", value);
	free(speed_class);
	
	char *uhs_speed_grade = grabHex(block_data_buff, 19, 1);
	if (uhs_speed_grade[0] == 0x00){
		sprintf(value, "UHS Speed Grade(0x13):\t\t\tLess than 10MB/s");
	}
	else if (uhs_speed_grade[0] == 0x01){
		sprintf(value, "UHS Speed Grade(0x13):\t\t\t10MB/s and above");
	}
	else if (uhs_speed_grade[0] == 0x03){
		sprintf(value, "UHS Speed Grade(0x13):\t\t\t30MB/s and above");
	}
	printf("\n%s", value);
	free(uhs_speed_grade);
	
	double rttb_count = hexArrToDec(block_data_buff, 26, 1);
	sprintf(value, "New Bad block Count(0x1A):\t\t%.0f", rttb_count);
	printf("\n%s", value);
	
	double spare_block = hexArrToDec(block_data_buff, 30, 2);
	sprintf(value, "Spare Block(0x1E):\t\t\t%.0f", spare_block);
	printf("\n%s", value);
	
	double min_erase_count_mt = hexArrToDec(block_data_buff, 32, 4);
	sprintf(value, "Min Erase Count(0x20):\t\t\t%.0f", min_erase_count_mt);
	printf("\n%s", value);
	
	double max_erase_count_mt = hexArrToDec(block_data_buff, 36, 4);
	sprintf(value, "Max Erase Count(0x24):\t\t\t%.0f", max_erase_count_mt);
	printf("\n%s", value);
	
	double total_erase_count_mt = hexArrToDec(block_data_buff, 40, 4);
	sprintf(value, "Total Erase Count(0x28):\t\t%.0f", total_erase_count_mt);
	printf("\n%s", value);
	
	double avg_erase_count_mt = hexArrToDec(block_data_buff, 44, 4);
	sprintf(value, "Avg. Erase Count(0x2C):\t\t\t%.0f", avg_erase_count_mt);
	printf("\n%s", value);
	
	double nand_pe_cycle = hexArrToDec(block_data_buff, 68, 2);
	sprintf(value, "NAND P/E Cycle(0x44):\t\t\t%.0f", nand_pe_cycle * 100);
	printf("\n%s", value);
	
	sprintf(value, "Card Life(0x46):\t\t\t%d %%", block_data_buff[70]);
	printf("\n%s", value);
	
	char *speed_mode = grabHex(block_data_buff, 71, 1);
	if (speed_mode[0] == 0x00){
		sprintf(value, "Current SD Card Speed Mode(0x47):\tDefault speed");
	}
	else if (speed_mode[0] == 0x01){
		sprintf(value, "Current SD Card Speed Mode(0x47):\tHigh speed mode");
	}
	else if (speed_mode[0] == 0x10){
		sprintf(value, "Current SD Card Speed Mode(0x47):\tSDR12");
	}
	else if (speed_mode[0] == 0x11){
		sprintf(value, "Current SD Card Speed Mode(0x47):\tSDR25");
	}
	else if (speed_mode[0] == 0x12){
		sprintf(value, "Current SD Card Speed Mode(0x47):\tSDR50");
	}
	else if (speed_mode[0] == 0x14){
		sprintf(value, "Current SD Card Speed Mode(0x47):\tDDR50");
	}
	else if (speed_mode[0] == 0x18){
		sprintf(value, "Current SD Card Speed Mode(0x47):\tSDR104");
	}
	printf("\n%s", value);
	free(speed_mode);
	
	double total_write_crc_count = hexArrToDec(block_data_buff, 72, 4);
	sprintf(value, "Total Write CRC Count(0x48):\t\t%.0f", total_write_crc_count);
	printf("\n%s", value);
	
	double power_on_off_count = hexArrToDec(block_data_buff, 76, 4);
	sprintf(value, "Power On/Off Count(0x4C):\t\t%.0f", power_on_off_count);
	printf("\n%s", value);

	char *nand_flash_id = grabHex(block_data_buff, 80, 6);
	sprintf(value, "NAND Flash ID(0x50):\t\t\t%02X-%02X-%02X-%02X-%02X-%02X", (unsigned char)nand_flash_id[0], (unsigned char)nand_flash_id[1], (unsigned char)nand_flash_id[2], (unsigned char)nand_flash_id[3], (unsigned char)nand_flash_id[4], (unsigned char)nand_flash_id[5]);
	printf("\n%s", value);
	free(nand_flash_id);
	
	
	char *smi_sd_controller_pn = grabString(block_data_buff, 88, 8);
	sprintf(value, "SMI SD Controller P/N(0x58):\t\t%s", smi_sd_controller_pn);
	printf("\n%s", value);
	free(smi_sd_controller_pn);
	
	char *sd_firmware_version = grabString(block_data_buff, 128, 6);
	sprintf(value, "SD Firmware Version(0x80):\t\t%s", sd_firmware_version);
	printf("\n%s\n", value);
	free(sd_firmware_version);
	
	return;
}

void parsing_Health_info(char *block_data_buff)
{
	char value[64];
	sprintf(value, "Card Life(0x46):\t\t\t%d %%", block_data_buff[70]);
	printf("\n%s\n", value);
	return;
}


char* grabString(char* data, int start_pos, int length){
	char *str = (char*)malloc( (length + 1) * sizeof(char) );
	strncpy(str, data + start_pos, length);
	str[length] = '\0';
	return str;
}

char* grabHex(char* data, int start_pos, int length){
	char *str = (char*)malloc( (length) * sizeof(char) );
	memcpy(str, data + start_pos, length);
	return str;
}

double hexArrToDec(char *data, int start_pos, int length){
	double tmp = 0;
	for(int i = 0; i < length; i++){
		double shift = (unsigned char)data[start_pos+i] << (4* (2 * ((length-i)-1))); 
		tmp += shift;
	}
	return tmp;
}


mmc-utils/mmc_cmds.h
/*
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License v2 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 021110-1307, USA.
 *
 * Modified to add field firmware update support,
 * those modifications are Copyright (c) 2016 SanDisk Corp.
 */

/* mmc_cmds.c */

int do_write_extcsd(int nargs, char **argv);
int do_writeprotect_boot_get(int nargs, char **argv);
int do_writeprotect_boot_set(int nargs, char **argv);
int do_writeprotect_user_get(int nargs, char **argv);
int do_writeprotect_user_set(int nargs, char **argv);
int do_disable_512B_emulation(int nargs, char **argv);
int do_write_boot_en(int nargs, char **argv);
int do_boot_bus_conditions_set(int nargs, char **argv);
int do_write_bkops_en(int nargs, char **argv);
int do_hwreset_en(int nargs, char **argv);
int do_hwreset_dis(int nargs, char **argv);
int do_sanitize(int nargs, char **argv);
int do_status_get(int nargs, char **argv);
int do_create_gp_partition(int nargs, char **argv);
int do_enh_area_set(int nargs, char **argv);
int do_write_reliability_set(int nargs, char **argv);
int do_rpmb_write_key(int nargs, char **argv);
int do_rpmb_read_counter(int nargs, char **argv);
int do_rpmb_read_block(int nargs, char **argv);
int do_rpmb_write_block(int nargs, char **argv);
int do_cache_en(int nargs, char **argv);
int do_cache_dis(int nargs, char **argv);
int do_ffu(int nargs, char **argv);
int do_read_scr(int argc, char **argv);
int do_read_cid(int argc, char **argv);
int do_read_csd(int argc, char **argv);

/* SD SMART */
int do_SMART_buffer_dump(int nargs, char **argv); //Show SMART raw buffer
int show_SMART_info(int nargs, char **argv); //Show SMART info
int show_Health_info(int nargs, char **argv); //Show Health info
int CMD56_data_in(int fd, int argCmd56, char *block_data_buff);
void dump_smart_data(char *block_data_buff);
void is_transcend_card(char *block_data_buff, char function);
void parsing_SMART_info(char *block_data_buff);
void parsing_Health_info(char *block_data_buff);
char *grabString(char* data, int start_pos, int length);
char *grabHex(char* data, int start_pos, int length);
double hexArrToDec(char *data, int start_pos, int length);


mmc-utils/mmc.c
/*
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License v2 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 021110-1307, USA.
 *
 * Modified to add field firmware update support,
 * those modifications are Copyright (c) 2016 SanDisk Corp.
 *
 * (This code is based on btrfs-progs/btrfs.c.)
 */

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "mmc_cmds.h"

#define MMC_VERSION	"0.1"

#define BASIC_HELP 0
#define ADVANCED_HELP 1

typedef int (*CommandFunction)(int argc, char **argv);

struct Command {
	CommandFunction	func;	/* function which implements the command */
	int	nargs;		/* if == 999, any number of arguments
				   if >= 0, number of arguments,
				   if < 0, _minimum_ number of arguments */
	char	*verb;		/* verb */
	char	*help;		/* help lines; from the 2nd line onward they 
                                   are automatically indented */
        char    *adv_help;      /* advanced help message; from the 2nd line 
                                   onward they are automatically indented */

	/* the following fields are run-time filled by the program */
	char	**cmds;		/* array of subcommands */
	int	ncmds;		/* number of subcommand */
};

static struct Command commands[] = {
	/*
	 *	avoid short commands different for the case only
	 */
	{ do_SMART_buffer_dump, -1,/* Get SMART buffer */
		"smtbuffer", "<device>\n",
		NULL
	},
	{ show_SMART_info, -1,     /* Get SMART info */
		"smart", "<device>\n",
		NULL
	},
	{ show_Health_info, -1,    /* Get Health info*/
		"health", "<device>\n",
		NULL
	},
	{ 0, 0, 0, 0 }
};

static char *get_prgname(char *programname)
{
	char	*np;
	np = strrchr(programname,'/');
	if(!np)
		np = programname;
	else
		np++;

	return np;
}

static void print_help(char *programname, struct Command *cmd, int helptype)
{
	char	*pc;

	printf("\t%s %s ", programname, cmd->verb );

	if (helptype == ADVANCED_HELP && cmd->adv_help)
		for(pc = cmd->adv_help; *pc; pc++){
			putchar(*pc);
			if(*pc == '\n')
				printf("\t\t");
		}
	else
		for(pc = cmd->help; *pc; pc++){
			putchar(*pc);
			if(*pc == '\n')
				printf("\t\t");
		}

	putchar('\n');
}

static void help(char *np)
{
	struct Command *cp;

	printf("Usage:\n");
	for( cp = commands; cp->verb; cp++ )
		print_help(np, cp, BASIC_HELP);

	printf("\n\t%s help|--help|-h\n\t\tShow the help.\n",np);
	printf("\n\t%s <cmd> --help\n\t\tShow detailed help for a command or subset of commands.\n",np);
	printf("\n%s\n", MMC_VERSION);
}

static int split_command(char *cmd, char ***commands)
{
	int	c, l;
	char	*p, *s;

	for( *commands = 0, l = c = 0, p = s = cmd ; ; p++, l++ ){
		if ( *p && *p != ' ' )
			continue;

		/* c + 2 so that we have room for the null */
		(*commands) = realloc( (*commands), sizeof(char *)*(c + 2));
		(*commands)[c] = strndup(s, l);
		c++;
		l = 0;
		s = p+1;
		if( !*p ) break;
	}

	(*commands)[c] = 0;
	return c;
}

/*
	This function checks if the passed command is ambiguous
*/
static int check_ambiguity(struct Command *cmd, char **argv){
	int		i;
	struct Command	*cp;
	/* check for ambiguity */
	for( i = 0 ; i < cmd->ncmds ; i++ ){
		int match;
		for( match = 0, cp = commands; cp->verb; cp++ ){
			int	j, skip;
			char	*s1, *s2;

			if( cp->ncmds < i )
				continue;

			for( skip = 0, j = 0 ; j < i ; j++ )
				if( strcmp(cmd->cmds[j], cp->cmds[j])){
					skip=1;
					break;
				}
			if(skip)
				continue;

			if( !strcmp(cmd->cmds[i], cp->cmds[i]))
				continue;
			for(s2 = cp->cmds[i], s1 = argv[i+1];
				*s1 == *s2 && *s1; s1++, s2++ ) ;
			if( !*s1 )
				match++;
		}
		if(match){
			int j;
			fprintf(stderr, "ERROR: in command '");
			for( j = 0 ; j <= i ; j++ )
				fprintf(stderr, "%s%s",j?" ":"", argv[j+1]);
			fprintf(stderr, "', '%s' is ambiguous\n",argv[j]);
			return -2;
		}
	}
	return 0;
}

/*
 * This function, compacts the program name and the command in the first
 * element of the '*av' array
 */
static int prepare_args(int *ac, char ***av, char *prgname, struct Command *cmd ){

	char	**ret;
	int	i;
	char	*newname;

	ret = (char **)malloc(sizeof(char*)*(*ac+1));
	newname = (char*)malloc(strlen(prgname)+strlen(cmd->verb)+2);
	if( !ret || !newname ){
		free(ret);
		free(newname);
		return -1;
	}

	ret[0] = newname;
	for(i=0; i < *ac ; i++ )
		ret[i+1] = (*av)[i];

	strcpy(newname, prgname);
	strcat(newname, " ");
	strcat(newname, cmd->verb);

	(*ac)++;
	*av = ret;

	return 0;

}

/*
	This function performs the following jobs:
	- show the help if '--help' or 'help' or '-h' are passed
	- verify that a command is not ambiguous, otherwise show which
	  part of the command is ambiguous
	- if after a (even partial) command there is '--help' show detailed help
	  for all the matching commands
	- if the command doesn't match show an error
	- finally, if a command matches, they return which command matched and
	  the arguments

	The function return 0 in case of help is requested; <0 in case
	of uncorrect command; >0 in case of matching commands
	argc, argv are the arg-counter and arg-vector (input)
	*nargs_ is the number of the arguments after the command (output)
	**cmd_  is the invoked command (output)
	***args_ are the arguments after the command

*/
static int parse_args(int argc, char **argv,
		      CommandFunction *func_,
		      int *nargs_, char **cmd_, char ***args_ )
{
	struct Command	*cp;
	struct Command	*matchcmd=0;
	char		*prgname = get_prgname(argv[0]);
	int		i=0, helprequested=0;

	if( argc < 2 || !strcmp(argv[1], "help") ||
		!strcmp(argv[1], "-h") || !strcmp(argv[1], "--help")){
		help(prgname);
		return 0;
	}

	for( cp = commands; cp->verb; cp++ )
		if( !cp->ncmds)
			cp->ncmds = split_command(cp->verb, &(cp->cmds));

	for( cp = commands; cp->verb; cp++ ){
		int     match;

		if( argc-1 < cp->ncmds )
			continue;
		for( match = 1, i = 0 ; i < cp->ncmds ; i++ ){
			char	*s1, *s2;
			s1 = cp->cmds[i];
			s2 = argv[i+1];

			for(s2 = cp->cmds[i], s1 = argv[i+1];
				*s1 == *s2 && *s1;
				s1++, s2++ ) ;
			if( *s1 ){
				match=0;
				break;
			}
		}

		/* If you understand why this code works ...
			you are a genious !! */
		if(argc>i+1 && !strcmp(argv[i+1],"--help")){
			if(!helprequested)
				printf("Usage:\n");
			print_help(prgname, cp, ADVANCED_HELP);
			helprequested=1;
			continue;
		}

		if(!match)
			continue;

		matchcmd = cp;
		*nargs_  = argc-matchcmd->ncmds-1;
		*cmd_ = matchcmd->verb;
		*args_ = argv+matchcmd->ncmds+1;
		*func_ = cp->func;

		break;
	}

	if(helprequested){
		printf("\n%s\n", MMC_VERSION);
		return 0;
	}

	if(!matchcmd){
		fprintf( stderr, "ERROR: unknown command '%s'\n",argv[1]);
		help(prgname);
		return -1;
	}

	if(check_ambiguity(matchcmd, argv))
		return -2;

	/* check the number of argument */
	if (matchcmd->nargs < 0 && matchcmd->nargs < -*nargs_ ){
		fprintf(stderr, "ERROR: '%s' requires minimum %d arg(s)\n",
			matchcmd->verb, -matchcmd->nargs);
			return -2;
	}
	if(matchcmd->nargs >= 0 && matchcmd->nargs != *nargs_ && matchcmd->nargs != 999){
		fprintf(stderr, "ERROR: '%s' requires %d arg(s)\n",
			matchcmd->verb, matchcmd->nargs);
			return -2;
	}
	
        if (prepare_args( nargs_, args_, prgname, matchcmd )){
                fprintf(stderr, "ERROR: not enough memory\\n");
		return -20;
        }


	return 1;
}
int main(int ac, char **av )
{
	char		*cmd=0, **args=0;
	int		nargs=0, r;
	CommandFunction func=0;

	r = parse_args(ac, av, &func, &nargs, &cmd, &args);
	if( r <= 0 ){
		/* error or no command to parse*/
		exit(-r);
	}

	exit(func(nargs, args));
}
